sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationRepository.java

```diff
@@ -0,0 +1,314 @@
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.identity.configuration;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.opensearch.identity.ConfigConstants;
+import org.opensearch.identity.exception.ConfigUpdateAlreadyInProgressException;
+import org.opensearch.identity.exception.ExceptionUtils;
+import org.opensearch.identity.exception.InvalidConfigException;
+import org.opensearch.identity.support.ConfigHelper;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.LogManager;
+import org.opensearch.OpenSearchException;
+import org.opensearch.ExceptionsHelper;
+import org.opensearch.ResourceAlreadyExistsException;
+import org.opensearch.action.admin.cluster.health.ClusterHealthRequest;
+import org.opensearch.action.admin.cluster.health.ClusterHealthResponse;
+import org.opensearch.action.admin.indices.create.CreateIndexRequest;
+import org.opensearch.client.Client;
+import org.opensearch.cluster.health.ClusterHealthStatus;
+import org.opensearch.cluster.metadata.IndexMetadata;
+import org.opensearch.cluster.metadata.MappingMetadata;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.common.util.concurrent.ThreadContext.StoredContext;
+import org.opensearch.env.Environment;
+import org.opensearch.threadpool.ThreadPool;
+
+public class ConfigurationRepository {
+    private static final Logger LOGGER = LogManager.getLogger(ConfigurationRepository.class);
+
+    private final String securityIndex;
+    private final Client client;
+    private final Map<CType, SecurityDynamicConfiguration<?>> configCache = new HashMap<>();
+    private final List<ConfigurationChangeListener> configurationChangedListener;
+    private final ConfigurationLoader cl;
+    private final Settings settings;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+    private DynamicConfigFactory dynamicConfigFactory;
+    private static final int DEFAULT_CONFIG_VERSION = 1;
+    private final Thread bgThread;
+    private final AtomicBoolean installDefaultConfig = new AtomicBoolean();
+
+    private ConfigurationRepository(
+        Settings settings,
+        final Path configPath,
+        ThreadPool threadPool,
+        Client client,
+        ClusterService clusterService
+    ) {
+        this.securityIndex = settings.get(ConfigConstants.IDENTITY_CONFIG_INDEX_NAME, ConfigConstants.IDENTITY_DEFAULT_CONFIG_INDEX);
```

Comment:
nit: we can rename this to identityIndex throughout this file

Reply:
Good idea

Reply:
Done